benchmarks/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,213 @@
+the performance of a numpy implementation with a python
+implementation.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+from datetime import datetime, timedelta, timezone
+from functools import partial
+from timeit import timeit
+from typing import List
+
+import numpy as np
+from frequenz.channels import Broadcast
+from numpy.random import default_rng
+from numpy.typing import NDArray
+
+from frequenz.sdk.timeseries import Sample
+from frequenz.sdk.timeseries._moving_window import MovingWindow
+from frequenz.sdk.timeseries.periodic_feature_extractor import PeriodicFeatureExtractor
+
+
+async def init_feature_extractor(period: int) -> PeriodicFeatureExtractor:
+    """Initialize the PeriodicFeatureExtractor class."""
+    # We only need the moving window to initialize the PeriodicFeatureExtractor class.
+    lm_chan = Broadcast[Sample]("lm_net_power")
+    # monkey patch the MovingWindow class to have a fixed size such that
+    # it can be initialized with a period bigger then the window size.
+    MovingWindow.__len__ = lambda self: period  # type: ignore[method-assign]
+    moving_window = MovingWindow(
+        timedelta(seconds=1), lm_chan.new_receiver(), timedelta(seconds=1)
+    )
+
+    await lm_chan.new_sender().send(Sample(datetime.now(tz=timezone.utc), 0))
+
+    # Initialize the PeriodicFeatureExtractor class with a period of period seconds.
+    # This works since the sampling period is set to 1 second.
+    return PeriodicFeatureExtractor(moving_window, timedelta(seconds=period))
+
+
+def _calculate_avg_window(
+    feature_extractor: PeriodicFeatureExtractor,
+    window: NDArray[np.float_],
+    window_size: int,
+) -> NDArray[np.float_]:
+    """
+    Reshapes the window and calculates the average.
+
+    This method calculates the average of a window by averaging over
+    all windows fully inside the passed numpy array having the distance
+    `self.distance`.
+
+    Args:
+        feature_extractor: The instance of the PeriodicFeatureExtractor to use.
+        window: The window to calculate the average over.
+        window_size: The size of the window to calculate the average over.
+        weights: The weights to use for the average calculation.
+
+    Returns:
+        The averaged window.
+    """
+    reshaped = feature_extractor._reshape_np_array(  # pylint: disable=W0212
+        window, window_size
+    )  # pylint: disable=W0212
+    # ignoring the type because np.average returns Any
+    return np.average(reshaped[:, :window_size], axis=0)  # type: ignore[no-any-return]
+
+
+def _calculate_avg_window_py(
+    feature_extractor: PeriodicFeatureExtractor,
+    window: NDArray[np.float_],
+    window_size: int,
+    weights: List[float] | None = None,
+) -> NDArray[np.float_]:
+    """
+    Plain python version of the average calculator.
+
+    This method avoids copying in any case but is 15 to 600
+    slower then the numpy version.
+
+    This method is only used in these benchmarks.
```

Comment:
There's an implementation of this method in the `PeriodicFeatureExtractor` class,  with the same code.  Can we drop this function in the benchmark file and use the one from the class?  If not,  maybe useful to document why not.

Reply:
Ups. That should not be in the `PeriodicFeatureExtractor` any longer. I'll will remove from there.